In [1]:
## Importing the libraries
import pandas as pd
from sklearn.linear_model import LinearRegression

from sklearn.model_selection import train_test_split, GridSearchCV
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error
from sklearn.impute import SimpleImputer
import numpy as np
from sklearn.tree import DecisionTreeRegressor
from sklearn.ensemble import RandomForestRegressor

In [2]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


# Loading the Training Data and using the Training Data to build the Model

In [3]:
df_train = pd.read_csv('/content/drive/My Drive/aa-project/Hackathon-2/Train_dataset.csv')
df_test = pd.read_csv('/content/drive/My Drive/aa-project/Hackathon-2/Test_dataset.csv')
sample_solution = pd.read_csv('/content/drive/My Drive/aa-project/Hackathon-2/Sample_Solution.csv')

In [4]:
df_train.head()

Registration Number  Annual Turnover           Cuisine       City  \
0                60001         42000000      indian,irish  Bangalore   
1                60002         50000000      indian,irish     Indore   
2                60003         32500000   tibetan,italian    Chennai   
3                60004        110000000  turkish,nigerian    Gurgaon   
4                60005         20000000     irish,belgian    Manesar   

  Restaurant Location Opening Day of Restaurant  Facebook Popularity Quotient  \
0   Near Business Hub                14-02-2009                          84.3   
1      Near Party Hub                29-09-2008                          85.4   
2   Near Business Hub                30-07-2011                          85.0   
3      Near Party Hub                30-11-2008                          85.6   
4      Near Party Hub                22-02-2010                           NaN   

        Endorsed By  Instagram Popularity Quotient  Fire Audit  ...  \
0      Not Specific                           95.8           1  ...   
1  Tier A Celebrity                           85.0           1  ...   
2  Tier A Celebrity                           68.2           1  ...   
3  Tier A Celebrity                           83.6           0  ...   
4  Tier A Celebrity                           76.8           1  ...   

   Overall Restaurant Rating  Live Music Rating  Comedy Gigs Rating  \
0                       10.0                4.0                 NaN   
1                        9.0                NaN                 4.0   
2                        8.0                3.0                 NaN   
3                        9.0                6.0                 NaN   
4                        6.0                NaN                 2.0   

   Value Deals Rating  Live Sports Rating Ambience Lively  Service  \
0                 NaN                 NaN      8.0      8        6   
1                 NaN                 NaN      5.0      7        7   
2                 NaN                 NaN      7.0     10        5   
3                 NaN                 NaN      7.0      7        4   
4                 NaN                 NaN      NaN      6        2   

   Comfortablility  Privacy  
0                6        6  
1                3        8  
2                2        8  
3                3        5  
4                4        6  

[5 rows x 34 columns]

In [5]:
df_train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3493 entries, 0 to 3492
Data columns (total 34 columns):
 #   Column                         Non-Null Count  Dtype  
---  ------                         --------------  -----  
 0   Registration Number            3493 non-null   int64  
 1   Annual Turnover                3493 non-null   int64  
 2   Cuisine                        3493 non-null   object 
 3   City                           3493 non-null   object 
 4   Restaurant Location            3493 non-null   object 
 5   Opening Day of Restaurant      3493 non-null   object 
 6   Facebook Popularity Quotient   3394 non-null   float64
 7   Endorsed By                    3493 non-null   object 
 8   Instagram Popularity Quotient  3437 non-null   float64
 9   Fire Audit                     3493 non-null   int64  
 10  Liquor License Obtained        3493 non-null   int64  
 11  Situated in a Multi Complex    3493 non-null   int64  
 12  Dedicated Parking              3493 non-null   i

In [6]:
df_test.head()

Registration Number           Cuisine       City Restaurant Location  \
0                20001   tibetan,italian  Bangalore   Near Business Hub   
1                20002   tibetan,italian  Hyderabad      Near Party Hub   
2                20003  algerian,belgian  Hyderabad      Near Party Hub   
3                20004     tibetan,greek         -1      Near Party Hub   
4                20005     cuban,british      Pune       Near Party Hub   

  Opening Day of Restaurant  Facebook Popularity Quotient      Endoresed By  \
0                13-07-2010                         78.00  Tier A Celebrity   
1                05-09-2011                         89.17      Not Specific   
2                12-04-2011                         84.00      Not Specific   
3                16-01-2005                         79.38      Not Specific   
4                10-11-2008                         84.67      Not Specific   

   Instagram Popularity Quotient  Fire Audit  Liquor License Obtained  ...  \
0                          69.00           1                        1  ...   
1                          96.00           1                        1  ...   
2                          86.00           1                        1  ...   
3                          74.40           0                        1  ...   
4                          86.46           1                        1  ...   

   Overall Restaurant Rating  Live Music Rating  Comedy Gigs Rating  \
0                        6.0                NaN                 2.0   
1                        9.0                3.0                 NaN   
2                        6.0                3.0                 NaN   
3                        9.0                6.0                 NaN   
4                        NaN                NaN                 NaN   

   Value Deals Rating Live Sports Rating Ambience  Lively  Service  \
0                 NaN                NaN        5       2        2   
1                 NaN                6.0        7       8        5   
2                 NaN                NaN        9       7        5   
3                 NaN                NaN        6       7        6   
4                 NaN                NaN        4       4        7   

   Comfortablility  Privacy  
0                6        0  
1                1        8  
2                1        5  
3                4        7  
4                7        3  

[5 rows x 33 columns]

In [7]:
sample_solution.head()

Registration Number  Annual Turnover
0                20001         25000000
1                20002         25000000
2                20003         25000000
3                20004         25000000
4                20005         25000000

In [8]:
df_train.isnull().sum().sort_values(ascending=False)

Live Sports Rating               3288
Value Deals Rating               2707
Comedy Gigs Rating               2483
Live Music Rating                 765
Overall Restaurant Rating         212
Facebook Popularity Quotient       99
Instagram Popularity Quotient      56
Resturant Tier                     49
Ambience                           25
Value for Money                     0
Hygiene Rating                      0
Food Rating                         0
Registration Number                 0
Order Wait Time                     0
Lively                              0
Service                             0
Comfortablility                     0
Staff Responsivness                 0
Restaurant Zomato Rating            0
Restaurant City Tier                0
Annual Turnover                     0
Restaurant Theme                    0
Restaurant Type                     0
Open Sitting Available              0
Dedicated Parking                   0
Situated in a Multi Complex         0
Liquor License Obtained             0
Fire Audit                          0
Endorsed By                         0
Opening Day of Restaurant           0
Restaurant Location                 0
City                                0
Cuisine                             0
Privacy                             0
dtype: int64

In [9]:
df_test.isnull().sum().sort_values(ascending=False)

Live Sports Rating               471
Value Deals Rating               385
Comedy Gigs Rating               370
Live Music Rating                102
Overall Restaurant Rating         34
Registration Number                0
Restaurant City Tier               0
Comfortablility                    0
Service                            0
Lively                             0
Ambience                           0
Food Rating                        0
Hygiene Rating                     0
Value for Money                    0
Staff Responsivness                0
Order Wait Time                    0
Restaurant Zomato Rating           0
Cuisine                            0
Restaurant Theme                   0
Restaurant Type                    0
Resturant Tier                     0
Open Sitting Available             0
Dedicated Parking                  0
Situated in a Multi Complex        0
Liquor License Obtained            0
Fire Audit                         0
Instagram Popularity Quotient      0
Endoresed By                       0
Facebook Popularity Quotient       0
Opening Day of Restaurant          0
Restaurant Location                0
City                               0
Privacy                            0
dtype: int64

In [10]:
# Exclude the 'Annual Turnover' column when fitting the imputer
numerical_features = df_train.select_dtypes(include=[np.number]).drop(columns=['Annual Turnover']).columns

# Fit the imputer on the training data
imputer = SimpleImputer(strategy='mean')
df_train_imputed = pd.DataFrame(imputer.fit_transform(df_train[numerical_features]), columns=numerical_features)


In [11]:
# Apply the imputer to the test data using the same features
df_test_imputed = pd.DataFrame(imputer.transform(df_test[numerical_features]), columns=numerical_features)

In [12]:
df_train_imputed.isnull().sum().sort_values(ascending=False)

Registration Number              0
Facebook Popularity Quotient     0
Comfortablility                  0
Service                          0
Lively                           0
Ambience                         0
Live Sports Rating               0
Value Deals Rating               0
Comedy Gigs Rating               0
Live Music Rating                0
Overall Restaurant Rating        0
Food Rating                      0
Hygiene Rating                   0
Value for Money                  0
Staff Responsivness              0
Order Wait Time                  0
Restaurant City Tier             0
Restaurant Zomato Rating         0
Resturant Tier                   0
Open Sitting Available           0
Dedicated Parking                0
Situated in a Multi Complex      0
Liquor License Obtained          0
Fire Audit                       0
Instagram Popularity Quotient    0
Privacy                          0
dtype: int64

In [13]:
df_test_imputed.isnull().sum().sort_values(ascending=False)

Registration Number              0
Facebook Popularity Quotient     0
Comfortablility                  0
Service                          0
Lively                           0
Ambience                         0
Live Sports Rating               0
Value Deals Rating               0
Comedy Gigs Rating               0
Live Music Rating                0
Overall Restaurant Rating        0
Food Rating                      0
Hygiene Rating                   0
Value for Money                  0
Staff Responsivness              0
Order Wait Time                  0
Restaurant City Tier             0
Restaurant Zomato Rating         0
Resturant Tier                   0
Open Sitting Available           0
Dedicated Parking                0
Situated in a Multi Complex      0
Liquor License Obtained          0
Fire Audit                       0
Instagram Popularity Quotient    0
Privacy                          0
dtype: int64

In [14]:
df_train_imputed.head()

Registration Number  Facebook Popularity Quotient  \
0              60001.0                     84.300000   
1              60002.0                     85.400000   
2              60003.0                     85.000000   
3              60004.0                     85.600000   
4              60005.0                     77.938715   

   Instagram Popularity Quotient  Fire Audit  Liquor License Obtained  \
0                           95.8         1.0                      1.0   
1                           85.0         1.0                      1.0   
2                           68.2         1.0                      1.0   
3                           83.6         0.0                      1.0   
4                           76.8         1.0                      1.0   

   Situated in a Multi Complex  Dedicated Parking  Open Sitting Available  \
0                          1.0                0.0                     1.0   
1                          1.0                1.0                     0.0   
2                          1.0                1.0                     0.0   
3                          1.0                1.0                     1.0   
4                          1.0                1.0                     1.0   

   Resturant Tier  Restaurant Zomato Rating  ...  Overall Restaurant Rating  \
0        2.000000                       3.0  ...                       10.0   
1        2.000000                       3.0  ...                        9.0   
2        2.000000                       3.0  ...                        8.0   
3        1.000000                       3.0  ...                        9.0   
4        1.926539                       3.0  ...                        6.0   

   Live Music Rating  Comedy Gigs Rating  Value Deals Rating  \
0            4.00000            2.932673            3.655216   
1            4.01283            4.000000            3.655216   
2            3.00000            2.932673            3.655216   
3            6.00000            2.932673            3.655216   
4            4.01283            2.000000            3.655216   

   Live Sports Rating  Ambience  Lively  Service  Comfortablility  Privacy  
0            3.590244   8.00000     8.0      6.0              6.0      6.0  
1            3.590244   5.00000     7.0      7.0              3.0      8.0  
2            3.590244   7.00000    10.0      5.0              2.0      8.0  
3            3.590244   7.00000     7.0      4.0              3.0      5.0  
4            3.590244   6.42301     6.0      2.0              4.0      6.0  

[5 rows x 26 columns]

In [15]:
df_train_imputed.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3493 entries, 0 to 3492
Data columns (total 26 columns):
 #   Column                         Non-Null Count  Dtype  
---  ------                         --------------  -----  
 0   Registration Number            3493 non-null   float64
 1   Facebook Popularity Quotient   3493 non-null   float64
 2   Instagram Popularity Quotient  3493 non-null   float64
 3   Fire Audit                     3493 non-null   float64
 4   Liquor License Obtained        3493 non-null   float64
 5   Situated in a Multi Complex    3493 non-null   float64
 6   Dedicated Parking              3493 non-null   float64
 7   Open Sitting Available         3493 non-null   float64
 8   Resturant Tier                 3493 non-null   float64
 9   Restaurant Zomato Rating       3493 non-null   float64
 10  Restaurant City Tier           3493 non-null   float64
 11  Order Wait Time                3493 non-null   float64
 12  Staff Responsivness            3493 non-null   f

In [16]:
df_test_imputed.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 500 entries, 0 to 499
Data columns (total 26 columns):
 #   Column                         Non-Null Count  Dtype  
---  ------                         --------------  -----  
 0   Registration Number            500 non-null    float64
 1   Facebook Popularity Quotient   500 non-null    float64
 2   Instagram Popularity Quotient  500 non-null    float64
 3   Fire Audit                     500 non-null    float64
 4   Liquor License Obtained        500 non-null    float64
 5   Situated in a Multi Complex    500 non-null    float64
 6   Dedicated Parking              500 non-null    float64
 7   Open Sitting Available         500 non-null    float64
 8   Resturant Tier                 500 non-null    float64
 9   Restaurant Zomato Rating       500 non-null    float64
 10  Restaurant City Tier           500 non-null    float64
 11  Order Wait Time                500 non-null    float64
 12  Staff Responsivness            500 non-null    flo

In [17]:
df_test_imputed.head()

Registration Number  Facebook Popularity Quotient  \
0              20001.0                         78.00   
1              20002.0                         89.17   
2              20003.0                         84.00   
3              20004.0                         79.38   
4              20005.0                         84.67   

   Instagram Popularity Quotient  Fire Audit  Liquor License Obtained  \
0                          69.00         1.0                      1.0   
1                          96.00         1.0                      1.0   
2                          86.00         1.0                      1.0   
3                          74.40         0.0                      1.0   
4                          86.46         1.0                      1.0   

   Situated in a Multi Complex  Dedicated Parking  Open Sitting Available  \
0                          1.0                1.0                     0.0   
1                          1.0                1.0                     1.0   
2                          1.0                0.0                     1.0   
3                          1.0                1.0                     1.0   
4                          1.0                0.0                     0.0   

   Resturant Tier  Restaurant Zomato Rating  ...  Overall Restaurant Rating  \
0             2.0                       3.0  ...                   6.000000   
1             2.0                       4.0  ...                   9.000000   
2             2.0                       2.0  ...                   6.000000   
3             2.0                       3.0  ...                   9.000000   
4             1.0                       4.0  ...                   8.479427   

   Live Music Rating  Comedy Gigs Rating  Value Deals Rating  \
0            4.01283            2.000000            3.655216   
1            3.00000            2.932673            3.655216   
2            3.00000            2.932673            3.655216   
3            6.00000            2.932673            3.655216   
4            4.01283            2.932673            3.655216   

   Live Sports Rating  Ambience  Lively  Service  Comfortablility  Privacy  
0            3.590244       5.0     2.0      2.0              6.0      0.0  
1            6.000000       7.0     8.0      5.0              1.0      8.0  
2            3.590244       9.0     7.0      5.0              1.0      5.0  
3            3.590244       6.0     7.0      6.0              4.0      7.0  
4            3.590244       4.0     4.0      7.0              7.0      3.0  

[5 rows x 26 columns]

# **Regression**

In [18]:
# Define the target variable and features
X = df_train_imputed
y = df_train['Annual Turnover']

# Split the data into training and validation sets
X_train, X_val, y_train, y_val = train_test_split(X, y, test_size=0.2, random_state=42)

# Initialize and train the Linear Regression model
model = LinearRegression()
model.fit(X_train, y_train)

# Predict on the test set
test_predictions = model.predict(df_test_imputed)


# **Loading the Test Data and using the Test Data to Predict the target Variable**

In [19]:
# Step 2: Merge predictions with the Sample Solution
solution_df = pd.DataFrame({
    'Registration Number': df_test['Registration Number'],
    'Predicted Annual Turnover': test_predictions
})

# Merge with the actual values from the Sample Solution
merged_df = pd.merge(solution_df, sample_solution, on='Registration Number')

# Step 3: Calculate RMSE
rmse = np.sqrt(mean_squared_error(merged_df['Annual Turnover'], merged_df['Predicted Annual Turnover']))
print(f"RMSE: {rmse}")

RMSE: 15174382.353311943


# Saving the output in a data frame and then exporting it to a '.csv' file with the appropriate 'Registration Number'.

In [20]:
import os
os.chdir('/content/drive/My Drive/aa-project/Hackathon-2')

In [21]:
# Assuming solution_df contains the 'Registration Number' and 'Predicted Annual Turnover' columns
solution_df = pd.DataFrame({
    'Registration Number': df_test['Registration Number'],
    'Annual Turnover': test_predictions
})

# Write the solution to a CSV file
solution_df.to_csv('Submission.csv', index=False)


# **Decision Tree**

In [22]:
# Initialize the Decision Tree Regressor
dt_regressor = DecisionTreeRegressor(random_state=42)

# Train the model on the training data
dt_regressor.fit(X_train, y_train)

# Predict on the validation set to evaluate the model
val_predictions = dt_regressor.predict(X_val)

# Calculate RMSE on the validation set
val_mse = mean_squared_error(y_val, val_predictions)
val_rmse = np.sqrt(val_mse)
print("Validation RMSE:", val_rmse)

# Predict on the test set
test_predictions = dt_regressor.predict(df_test_imputed)

# Display the test predictions
print(test_predictions)

Validation RMSE: 25727702.542426676
[3.80e+07 5.00e+07 1.80e+07 3.05e+07 1.10e+08 2.00e+07 3.30e+07 4.35e+07
 4.00e+07 3.40e+07 3.00e+07 5.25e+07 1.45e+07 1.50e+07 3.15e+07 3.25e+07
 3.60e+07 1.05e+07 4.50e+07 2.75e+07 2.20e+07 4.25e+07 1.20e+07 2.75e+07
 6.00e+07 1.80e+07 4.60e+07 3.65e+07 5.10e+07 2.90e+07 7.50e+07 3.25e+07
 3.80e+07 3.80e+07 1.20e+07 1.20e+07 5.10e+07 2.00e+07 1.30e+07 1.10e+07
 6.00e+07 5.00e+07 9.50e+06 2.20e+07 3.80e+07 1.20e+07 4.30e+07 2.75e+07
 3.10e+07 1.45e+07 3.90e+07 5.10e+07 1.85e+07 3.10e+07 3.00e+07 1.20e+07
 1.80e+07 4.20e+07 1.80e+07 3.50e+07 2.20e+07 6.00e+07 3.15e+07 2.40e+07
 2.00e+07 2.50e+07 2.40e+07 2.00e+07 2.10e+07 3.10e+07 3.05e+07 1.85e+07
 2.40e+07 4.60e+07 3.00e+07 3.35e+07 3.35e+07 4.80e+07 1.25e+07 1.20e+08
 3.25e+07 4.45e+07 1.80e+07 5.00e+07 3.50e+07 3.00e+07 1.00e+07 2.00e+07
 5.60e+07 3.00e+07 5.50e+07 1.80e+07 3.80e+07 2.00e+07 3.20e+07 2.75e+07
 1.80e+07 1.45e+07 6.20e+07 3.15e+07 3.35e+07 1.80e+07 4.00e+07 3.15e+07
 3.80e+07 1.10e

# **Pre-Prunning (Decision Tree)**

In [23]:
# Define the Decision Tree Regressor
dt_regressor = DecisionTreeRegressor(random_state=42)

# Define the parameter grid for Grid Search
param_grid = {
    'max_depth': [None, 10, 20, 30, 40, 50],
    'min_samples_split': [2, 5, 10],
    'min_samples_leaf': [1, 2, 4]
}

# Set up Grid Search with cross-validation
grid_search = GridSearchCV(estimator=dt_regressor,
                           param_grid=param_grid,
                           scoring='neg_mean_squared_error',
                           cv=5,  # 5-fold cross-validation
                           n_jobs=-1,
                           verbose=1)

# Fit Grid Search to the training data
grid_search.fit(X_train, y_train)

# Get the best parameters and best model
best_params = grid_search.best_params_
best_model = grid_search.best_estimator_

# Print the best parameters
print("Best Parameters:", best_params)

# Predict on the validation set
val_predictions = best_model.predict(X_val)

# Calculate RMSE on the validation set
val_mse = mean_squared_error(y_val, val_predictions)
val_rmse = np.sqrt(val_mse)
print("Validation RMSE:", val_rmse)

# Predict on the test set
test_predictions = best_model.predict(df_test_imputed)

# Display the test predictions
print(test_predictions)

Fitting 5 folds for each of 54 candidates, totalling 270 fits
Best Parameters: {'max_depth': 10, 'min_samples_leaf': 1, 'min_samples_split': 10}
Validation RMSE: 22736319.5665186
[2.74583333e+07 4.71111111e+07 1.90000000e+07 3.19473684e+07
 1.10000000e+08 1.66666667e+07 3.97500000e+07 3.25416667e+07
 2.90280899e+07 2.18125000e+07 3.50357143e+07 5.25000000e+07
 1.82777778e+07 2.17933673e+07 2.77857143e+07 2.90280899e+07
 3.19473684e+07 2.37806604e+07 4.29508197e+07 3.10512821e+07
 2.17933673e+07 4.29508197e+07 2.18125000e+07 3.10512821e+07
 3.46363636e+07 2.17933673e+07 3.50357143e+07 3.45454545e+07
 3.97500000e+07 2.28431373e+07 7.50000000e+07 3.51153846e+07
 3.46666667e+07 3.46666667e+07 2.90280899e+07 2.37806604e+07
 3.97500000e+07 1.61250000e+07 2.48357143e+07 2.17933673e+07
 3.46363636e+07 4.75000000e+07 1.55000000e+07 2.18125000e+07
 3.46666667e+07 2.37806604e+07 2.18125000e+07 3.50357143e+07
 2.77857143e+07 2.18125000e+07 4.29508197e+07 3.97500000e+07
 2.17933673e+07 2.79166667e+

# **Post-Pruning (Decision Tree)**

In [24]:
# Train a full Decision Tree Regressor
full_tree = DecisionTreeRegressor(random_state=42)
full_tree.fit(X_train, y_train)

# Calculate the complexity pruning path
path = full_tree.cost_complexity_pruning_path(X_train, y_train)
alphas = path.ccp_alphas
impurities = path.impurities

# Train Decision Trees for each alpha
trees = []
for alpha in alphas:
    tree = DecisionTreeRegressor(random_state=42, ccp_alpha=alpha)
    tree.fit(X_train, y_train)
    trees.append(tree)

# Evaluate each tree on the validation set
val_rmse = []
for tree in trees:
    val_predictions = tree.predict(X_val)
    val_mse = mean_squared_error(y_val, val_predictions)
    val_rmse.append(np.sqrt(val_mse))

# Find the index of the best tree
best_index = np.argmin(val_rmse)
best_alpha = alphas[best_index]
best_tree = trees[best_index]

print("Best alpha:", best_alpha)
print("Validation RMSE with best alpha:", val_rmse[best_index])

# Predict on the test set with the best tree
test_predictions = best_tree.predict(df_test_imputed)

# Display the test predictions
print(test_predictions)

Best alpha: 4945084644606.479
Validation RMSE with best alpha: 20568399.6723921
[29565217.39130435 35573321.55477032 35573321.55477032 39051369.8630137
 41651785.71428572 35573321.55477032 35573321.55477032 25264363.03080766
 25264363.03080766 26289915.96638655 29565217.39130435 35573321.55477032
 35573321.55477032 25264363.03080766 29565217.39130435 25264363.03080766
 39051369.8630137  25264363.03080766 39051369.8630137  26289915.96638655
 25264363.03080766 39051369.8630137  26289915.96638655 26289915.96638655
 34468152.86624204 25264363.03080766 29565217.39130435 35573321.55477032
 35573321.55477032 25264363.03080766 35573321.55477032 35573321.55477032
 29565217.39130435 29565217.39130435 25264363.03080766 25264363.03080766
 35573321.55477032 35573321.55477032 25264363.03080766 25264363.03080766
 34468152.86624204 26289915.96638655 29565217.39130435 26289915.96638655
 29565217.39130435 25264363.03080766 26289915.96638655 29565217.39130435
 29565217.39130435 26289915.96638655 39051369

# **Random Forest**

In [25]:
# Initialize the Random Forest Regressor
rf_regressor = RandomForestRegressor(random_state=42, n_estimators=100)

# Train the model on the training data
rf_regressor.fit(X_train, y_train)

# Predict on the validation set
val_predictions = rf_regressor.predict(X_val)

# Calculate RMSE on the validation set
val_mse = mean_squared_error(y_val, val_predictions)
val_rmse = np.sqrt(val_mse)
print("Validation RMSE:", val_rmse)

# Predict on the test set
test_predictions = rf_regressor.predict(df_test_imputed)

# Display the test predictions
print(test_predictions)

Validation RMSE: 20455733.701660458
[32005000. 47800000. 33495000. 39140000. 67595000. 43915000. 33925000.
 23245000. 32820000. 25235000. 39720000. 36180000. 40395000. 26855000.
 38150000. 33630000. 37975000. 21095000. 43645000. 23335000. 26275000.
 40255000. 23110000. 30605000. 35995000. 21750000. 42265000. 45020000.
 43685000. 26040000. 46870000. 48770000. 34170000. 34105000. 31000000.
 20025000. 43250000. 39605000. 28005000. 22290000. 31455000. 27235000.
 33020000. 21940000. 30635000. 21215000. 19725000. 35260000. 28455000.
 35265000. 41765000. 42895000. 20290000. 36585000. 31380000. 30530000.
 31745000. 38565000. 45460000. 33360000. 18280000. 37490000. 35840000.
 44815000. 37160000. 41360000. 24620000. 27010000. 27075000. 49740000.
 44610000. 22020000. 24495000. 20150000. 35455000. 46270000. 40880000.
 26435000. 25785000. 21400000. 42890000. 25820000. 44705000. 42040000.
 35865000. 32815000. 22300000. 39525000. 40070000. 34015000. 43665000.
 18605000. 28470000. 18875000. 46300000. 

In [26]:
import pandas as pd

# Create a dictionary with the results
results = {
    'Model': ['Linear Regression', 'Decision Tree', 'Pre-Pruning (Decision Tree)', 'Post-Pruning (Decision Tree)', 'Random Forest'],
    'RMSE': [15174382.353311943, 25727702.542426676, 22736319.5665186, 20568399.6723921, 20455733.701660458]
}

# Convert the dictionary to a DataFrame
results_df = pd.DataFrame(results)

# Set display format to show 2 decimal places
pd.options.display.float_format = '{:.2f}'.format

# Display the DataFrame
print(results_df)


                          Model        RMSE
0             Linear Regression 15174382.35
1                 Decision Tree 25727702.54
2   Pre-Pruning (Decision Tree) 22736319.57
3  Post-Pruning (Decision Tree) 20568399.67
4                 Random Forest 20455733.70


The Linear Regression model, with an RMSE of 15174382.35, has the lowest error compared to all other models evaluated.